### ISSUE TITLE
GUI

### ISSUE THREAD

**Issue by janbaudisch** [skycocker/chromebrew#889]
This issue is intended to keep track of the progress made in the case of GUI applications. It is a result of #886.

- [x] libX11 (#930)
- [x] X-org protocols (#906)
- [x] Wayland (#961)
- [x] Mesa (#962)
- [x] udev / eudev (#1228)
- [x] XWayland (#1229)
- [x] GTK
- [x] Sommelier

**Comment by smerkousdavid** [comment]
I think that's an amazing idea. Here are a few packages/dependencies for X11. 
https://github.com/CodeBuds/Crevos/tree/32c0200b6b2bc08152b31d1d952c58e0875748a6/packages.

Cheers

**Comment by bfayers** [comment]
Super cool, once this is working I may revert one of my chromebooks to running chromeos again instead of Arch.

**Comment by uberhacker** [comment]
How is this coming along @flyingP0tat0?  You making any progress?

**Comment by flyingP0tat0** [comment]
I don't really have time to finish this project unfortunately.
Feel free to check out [my fork](https://github.com/flyingP0tat0/chromebrew), especially the test directory. There already are Weston and GTK scripts. I did not get these working, but it is possible.
Although I don't really have time at the moment, you can tag me everytime you want me to take a look at something.

**Comment by uberhacker** [comment]
Yes, I understand.  This is a very ambitious project and I appreciate the effort you have put into it so far.  I'll see if I have time to nudge this along when I get a chance.
Thanks

**Comment by littlej247** [comment]
The work you guys are doing is absolutely amazing. I know this isn't the place for this but I didn't know where else to say it. Thanks for all your hard work to everyone involved in chromebrew. Honestly Thank You!

**Comment by nskaggs** [comment]
@flyingP0tat0, can you summarize what's left to make this a reality? Where exactly did you get stuck?

**Comment by flyingP0tat0** [comment]
@nskaggs I guess the most important thing is to get GTK and Weston to work.

**Comment by VitaliyKulikov** [comment]
ping me if need to test something on Google Pixelbook )) meantime, i will take a look on codebase.

**Comment by beliefless-dogmasless** [comment]
I'm linking this:
https://www.reddit.com/r/unixporn/comments/7zns09/chromebook_wayland_going_full_native/

**Comment by bfayers** [comment]
Whoah, cool!

**Comment by tomas** [comment]
@uberhacker Do you think it would be possible to get something working, now that the gtk3 package has been merged?

**Comment by flyingP0tat0** [comment]
What @beliefless-dogmasless linked was achieved with crouton. So that's unfortunately not the way to go for us.

**Comment by beliefless-dogmasless** [comment]
@flyingP0tat0 
Yes, that what I saw later, when 'the poster', made some clarifications.

**Comment by uberhacker** [comment]
Hey guys, I think we're definitely getting closer to the end goal but there is quite a bit more to do.  @lovedheart has made huge strides but we still have a way to go yet.  I'm not sure what the best approach is going forward.  There are several different ways to do it besides just the X Window System way which is a bit outdated.

**Comment by lovedheart** [comment]
I was able to run gtk3-demo when I fully complied the source code. But it's not working with pre-built binaries.

**Comment by uberhacker** [comment]
@lovedheart: A lot of the pre-built binaries have `--without-x` or `--without-gtk` configure options so you might try recompiling with those enabled.

**Comment by beliefless-dogmasless** [comment]
I agree gtk3-demo, works
<a href="http://fr.tinypic.com?ref=35by3v7" target="_blank"><img src="http://i67.tinypic.com/35by3v7.png" border="0" alt="Image and video hosting by TinyPic"></a>

There is 'a problem with the icons'

- I installed geany-gtk3, make a lot of time to start the first time, then it starts automatically(no time/no latence)

**Comment by beliefless-dogmasless** [comment]
Now there is 'works' to do, integrate in the shell(chromeos shell), make the icons(seeing them, and close/minimize buttons) works. The ability to control the windows with 'the finger', drag/and move.

**Comment by lovedheart** [comment]
@beliefless-dogmasless What did you do? Have you compiled gtk3 by yourself? Or just install the binary?

**Comment by beliefless-dogmasless** [comment]
I used the binary, I think(not sure).

**Comment by beliefless-dogmasless** [comment]
@lovedheart
I used the binary.(I just checked on the notes of installation I keep)

**Comment by lovedheart** [comment]
@beliefless-dogmasless  Could you share the notes?

**Comment by beliefless-dogmasless** [comment]
@lovedheart 
I made a fresh install, put me in 'canary mode/channel'
Otherwise nothing much, I put the notes here:
https://pastebin.com/U81WUaxW

PS:  I'm on Arm.

**Comment by gt945** [comment]
good job!!
Weston is working now
I use below config:
~/.config/weston.ini
```
[core]
xwayland=true
backend=wayland-backend.so
shell=desktop-shell.so
[xwayland]
path=/usr/local/bin/Xwayland
```

then run 
`weston`
or
`weston --fullscreen`

**Comment by uberhacker** [comment]
@gt945: Awesome!  Is that a standard config we can add to the `weston.rb` package?

**Comment by beliefless-dogmasless** [comment]
Don't forget xwayland:
https://wiki.archlinux.org/index.php/Wayland#Weston

[core]
modules=xwayland.so

**Comment by gt945** [comment]
@beliefless-dogmasless 
xwayland=true is recommended method now

**Comment by gt945** [comment]
@uberhacker  on chromebook, yes.
But the gl render not work, no idea yet.

**Comment by lzmartinico** [comment]
I just tried a fresh weston install with the recommended config, but starting it I get the following error message:
[16:58:09.875] weston 4.0.0
               http://wayland.freedesktop.org
               Bug reports to: https://bugs.freedesktop.org/enter_bug.cgi?product=Wayland&component=weston&version=4.0.0
               Build: unknown (not built from git or tarball)
[16:58:09.877] Command line: weston
[16:58:09.877] OS: Linux, 4.4.79-11603-g3448c40b1720, #1 SMP PREEMPT Fri Sep 15 00:46:44 UTC 2017, x86_64
[16:58:09.878] Using config file '/home/chronos/user/.config/weston.ini'
[16:58:09.881] Output repaint window is 7 ms maximum.
[16:58:09.882] Loading module '/usr/local/lib64/libweston-4/wayland-backend.so'
[16:58:09.923] Error: Failed to connect to parent Wayland compositor: No such file or directory
               display option: (none), WAYLAND_DISPLAY=(not set)
[16:58:09.923] fatal: failed to create compositor 
The WAYLAND_DISPLAY variable is unset; how did you manage to run it without the init daemon?

**Comment by gt945** [comment]
@lzmartinico There is a weird issue on my side,  weston only works after exec Xwayland one time.
can you check whether Xwayland works?
If work , quit Xwayland and run weston

**Comment by lzmartinico** [comment]
@gt945 thank you, but on a fresh install it's now running without having to run Xwayland first 

On July 9, 2018 3:01:54 AM UTC, Tao Guo <notifications@github.com> wrote:
>@lzmartinico There is a weird issue on my side,  weston only works
>after exec Xwayland one time.
>can you check whether Xwayland works?
>If work , quit Xwayland and run weston
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/skycocker/chromebrew/issues/889#issuecomment-403346621

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by superloach** [comment]
What is the overall status of this project? (eg, what can we get running so far?)

Also, Weston fails to build for me on an Acer R11 (CYAN), apparently due to a missing `libicui18n.so.60`. All the other previously listed components have been installed, as far as I'm aware. Full output: https://pastebin.com/JvWcFU6r

Just excited to see where we can take this, and willing to contribute my limited knowledge, and my machine for testing. 😃

**Comment by JL2210** [comment]
All of these are in, working on gnome now.

**Comment by superloach** [comment]
Awesome, can't wait for gnome. Are we thinking of getting Qt as well?

**Comment by JL2210** [comment]
Apparently already has qt

**Comment by JL2210** [comment]
@uberhacker, all the deps are satiafied for this. Can be closed.

**Comment by uberhacker** [comment]
I'll close this when it actually works. :)  Still need to figure out why code stopped working after 1.25.

**Comment by JL2210** [comment]
I installed XFCE and it works in that.

**Comment by uberhacker** [comment]
Works in what?

**Comment by JL2210** [comment]
Code works in XFCE.

**Comment by JL2210** [comment]
The desktop environment.

**Comment by uberhacker** [comment]
Well, where is the xfce PR then? ;)

**Comment by JL2210** [comment]
It wouldn't just be one PR.

**Comment by JL2210** [comment]
Look at the Tilp one I made, though.

**Comment by uberhacker** [comment]
If code works in xfce then maybe there are still some core dependencies missing for other environments such as wayland or weston.

**Comment by JL2210** [comment]
I have a hunch that it's sommelier.

**Comment by uberhacker** [comment]
Well, if code works in xfce with sommelier, why doesn't it work w/o xfce?

**Comment by JL2210** [comment]
Without sommelier. XFCE/Sommelier do not run with each other.

**Comment by JL2210** [comment]
XFCE is a proper window manager and session manager. Sommelier just forwards X to Wayland.

**Comment by uberhacker** [comment]
I get that but it still doesn't explain why code runs in xfce and not without.  Can you get a code version > 1.25 working in xfce?  If so, then there are some dependences for xfce that need added so code can work outside of xfce.

**Comment by JL2210** [comment]
Yes. No.

**Comment by uberhacker** [comment]
What does that mean?  Intelligent responses only, please. :)

**Comment by JL2210** [comment]
We could always xinit with Xwayland directly into code. No minimize/maximize/close bar, though.

**Comment by JL2210** [comment]
XFCE works with versions greater than 1.25. No extra dependencies.

**Comment by uberhacker** [comment]
Sounds more complicated than necessary.

**Comment by JL2210** [comment]
Use version 1.15, then.

**Comment by uberhacker** [comment]
Dude, I don't think you are understanding what I'm asking here.  Forget about xfce for a moment.  What are the dependency requirements for xfce?  One of those dependencies are likely the ticket to getting code working outside of xfce.

**Comment by JL2210** [comment]
They're all in the XFCE fat tarball.

**Comment by JL2210** [comment]
They're all in the XFCE fat tarball.

**Comment by JL2210** [comment]
```
```
	linux-vdso.so.1 (0x00007ffdbd35a000)
	libxfce4ui-1.so.0 => /usr/local/lib64/libxfce4ui-1.so.0 (0x00007bf0ba4fb000)
	libwnck-1.so.22 => /usr/local/lib64/libwnck-1.so.22 (0x00007bf0ba2bd000)
	libXRes.so.1 => /usr/local/lib64/libXRes.so.1 (0x00007bf0ba0b9000)
	libxfconf-0.so.2 => /usr/local/lib64/libxfconf-0.so.2 (0x00007bf0b9ea6000)
	libexo-1.so.0 => /usr/local/lib64/libexo-1.so.0 (0x00007bf0b9c5a000)
	libSM.so.6 => /usr/local/lib64/libSM.so.6 (0x00007bf0b9a53000)
	libICE.so.6 => /usr/local/lib64/libICE.so.6 (0x00007bf0b9837000)
	libX11.so.6 => /usr/lib64/libX11.so.6 (0x00007bf0ba7f8000)
	libgtk-x11-2.0.so.0 => /usr/local/lib64/libgtk-x11-2.0.so.0 (0x00007bf0b91fa000)
	libgdk-x11-2.0.so.0 => /usr/local/lib64/libgdk-x11-2.0.so.0 (0x00007bf0b8f47000)
	libpangocairo-1.0.so.0 => /usr/local/lib64/libpangocairo-1.0.so.0 (0x00007bf0b8d3a000)
	libatk-1.0.so.0 => /usr/local/lib64/libatk-1.0.so.0 (0x00007bf0b8b14000)
	libcairo.so.2 => /usr/local/lib64/libcairo.so.2 (0x00007bf0b87f6000)
	libgdk_pixbuf-2.0.so.0 => /usr/local/lib64/libgdk_pixbuf-2.0.so.0 (0x00007bf0b85d3000)
	libpangoft2-1.0.so.0 => /usr/local/lib64/libpangoft2-1.0.so.0 (0x00007bf0b83bf000)
	libpango-1.0.so.0 => /usr/local/lib64/libpango-1.0.so.0 (0x00007bf0b817b000)
	libfontconfig.so.1 => /usr/local/lib64/libfontconfig.so.1 (0x00007bf0b7f39000)
	libfreetype.so.6 => /usr/local/lib64/libfreetype.so.6 (0x00007bf0b7e53000)
	libm.so.6 => /usr/local/lib64/libm.so.6 (0x00007bf0b7b4d000)
	libgarcon-1.so.0 => /usr/local/lib64/libgarcon-1.so.0 (0x00007bf0b792f000)
	libxfce4util.so.7 => /usr/local/lib64/libxfce4util.so.7 (0x00007bf0b7721000)
	libgio-2.0.so.0 => /usr/local/lib64/libgio-2.0.so.0 (0x00007bf0b7385000)
	libgthread-2.0.so.0 => /usr/local/lib64/libgthread-2.0.so.0 (0x00007bf0b7183000)
	libdbus-glib-1.so.2 => /usr/local/lib64/libdbus-glib-1.so.2 (0x00007bf0ba7ca000)
	libdbus-1.so.3 => /usr/local/lib64/libdbus-1.so.3 (0x00007bf0b6f22000)
	libgobject-2.0.so.0 => /usr/local/lib64/libgobject-2.0.so.0 (0x00007bf0b6ccf000)
	libglib-2.0.so.0 => /usr/local/lib64/libglib-2.0.so.0 (0x00007bf0b698d000)
	libpthread.so.0 => /usr/local/lib64/libpthread.so.0 (0x00007bf0b6770000)
	libc.so.6 => /usr/local/lib64/libc.so.6 (0x00007bf0b63c5000)
	libXext.so.6 => /usr/local/lib64/libXext.so.6 (0x00007bf0b61b3000)
	libbsd.so.0 => /usr/local/lib64/libbsd.so.0 (0x00007bf0b5f9e000)
	librt.so.1 => /usr/local/lib64/librt.so.1 (0x00007bf0b5d96000)
	libuuid.so.1 => /usr/local/lib64/libuuid.so.1 (0x00007bf0b5b92000)
	libxcb.so.1 => /usr/lib64/libxcb.so.1 (0x00007bf0ba7a3000)
	libdl.so.2 => /usr/local/lib64/libdl.so.2 (0x00007bf0b598e000)
	libXrender.so.1 => /usr/local/lib64/libXrender.so.1 (0x00007bf0b5784000)
	libXinerama.so.1 => /usr/local/lib64/libXinerama.so.1 (0x00007bf0b5582000)
	libXrandr.so.2 => /usr/local/lib64/libXrandr.so.2 (0x00007bf0b5377000)
	libXcursor.so.1 => /usr/local/lib64/libXcursor.so.1 (0x00007bf0b516c000)
	libgmodule-2.0.so.0 => /usr/local/lib64/libgmodule-2.0.so.0 (0x00007bf0b4f68000)
	libXcomposite.so.1 => /usr/local/lib64/libXcomposite.so.1 (0x00007bf0b4d65000)
	libXdamage.so.1 => /usr/local/lib64/libXdamage.so.1 (0x00007bf0b4b62000)
	libXfixes.so.3 => /usr/local/lib64/libXfixes.so.3 (0x00007bf0b495c000)
	libfribidi.so.0 => /usr/local/lib64/libfribidi.so.0 (0x00007bf0b4741000)
	libharfbuzz.so.0 => /usr/lib64/libharfbuzz.so.0 (0x00007bf0b46b4000)
	libpixman-1.so.0 => /usr/local/lib64/libpixman-1.so.0 (0x00007bf0b440d000)
	libEGL.so.1 => /usr/lib64/libEGL.so.1 (0x00007bf0ba77f000)
	libGLESv2.so.2 => /usr/lib64/libGLESv2.so.2 (0x00007bf0ba775000)
	libpng16.so.16 => /usr/local/lib64/libpng16.so.16 (0x00007bf0b41db000)
	libxcb-shm.so.0 => /usr/lib64/libxcb-shm.so.0 (0x00007bf0ba770000)
	libX11-xcb.so.1 => /usr/lib64/libX11-xcb.so.1 (0x00007bf0ba76d000)
	libxcb-render.so.0 => /usr/lib64/libxcb-render.so.0 (0x00007bf0ba760000)
	libz.so.1 => /usr/local/lib64/libz.so.1 (0x00007bf0b3fbe000)
	libresolv.so.2 => /usr/local/lib64/../lib64/libresolv.so.2 (0x00007bf0b3da6000)
	libmount.so.1 => /usr/local/lib64/../lib64/libmount.so.1 (0x00007bf0b3b55000)
	libblkid.so.1 => /usr/local/lib64/../lib64/libblkid.so.1 (0x00007bf0b390c000)
	libffi.so.6 => /usr/local/lib64/../lib64/libffi.so.6 (0x00007bf0b3702000)
	libbz2.so.1 => /usr/local/lib64/libbz2.so.1 (0x00007bf0ba74a000)
	libexpat.so.1 => /usr/local/lib64/libexpat.so.1 (0x00007bf0b36ce000)
	/lib64/ld-linux-x86-64.so.2 (0x00007bf0ba70d000)
	libXau.so.6 => /usr/lib64/libXau.so.6 (0x00007bf0ba744000)
	libXdmcp.so.6 => /usr/lib64/libXdmcp.so.6 (0x00007bf0ba73d000)
	libdrm.so.2 => /usr/lib64/libdrm.so.2 (0x00007bf0b36bc000)
	libglapi.so.0 => /usr/lib64/libglapi.so.0 (0x00007bf0b3694000)
```

**Comment by JL2210** [comment]
Those are the requirements of xfdesktop. Try `garcon`.

**Comment by JL2210** [comment]
If I have XFCE installed, and it doesn't work in sommelier then, why do you expect any of these libs would make it work, even without XFCE?

**Comment by JL2210** [comment]
Specific error:
```
(code:18428): Gtk-WARNING **: 17:40:15.891: cannot open display: :0
```
(I know, not very helpful, but at least something to go on)

**Comment by willardthor** [comment]
@flyingP0tat0 @uberhacker is there anything that us users of chromebrew can do to help this along?

**Comment by JL2210** [comment]
`code` works now, I'll update it. I think this may finally be able to be closed.

**Comment by JL2210** [comment]
@uberhacker Can this be closed? Everything pretty much works, and like you said, any new issues can be created on the issues page instead of being pulled up here.

**Comment by uberhacker** [comment]
The code package is still not working for me and several other users.  I will leave this open until we can make sure all apps that use sommelier are working.

**Comment by JL2210** [comment]
The other apps that use Sommelier are working and Code is a lost cause.

**Comment by saltedcoffii** [comment]
Considering that this issue is so old, and that we have full support for sommelier on stable machines with x86_64 (amd64 and em64t) aarch64 and armv7l, I'm going to close this issue. Should anyone like it opened again, please ping me.
